File: list/Solution19.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution19 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
    	if(head == null || head.next == null){
    		return null;
    	}
      ListNode sentinel = new ListNode(1);
      sentinel.next = head;
      while(sentinel.next != null){
      	ListNode cur = sentinel.next;
      	int index = n;
      	while(index > 0){
      		cur = cur.next;
      		index -= 1;
      	}
      	if(cur == null){
      		if(sentinel.next == head) {
				sentinel.next = sentinel.next.next;
				return sentinel.next;
			}
      		else{
      			sentinel.next = sentinel.next.next;
      			return head;
			}

      	}
      	sentinel = sentinel.next;
      }
      return null;
    }

    public static void main(String[] args){
    	Solution19 s = new Solution19();
    	ListNode test = new ListNode(1);
    	ListNode next = new ListNode(2);
    	test.next = next;
    	ListNode result = s.removeNthFromEnd(test, 2);
    }
}

File: C++/Solution19.cpp
#include<iostream>
#include<vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution19 {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* pre = new ListNode(0, head);
        ListNode* ptr = pre;
        while(ptr -> next){
            ListNode* ptr2 = ptr;
            int copy = n;
            while(copy > 0){
                ptr2 = ptr2 -> next;
                copy--;
            }
            if(!ptr2 -> next){
                ptr -> next = ptr -> next -> next;
                return pre -> next;
            }
            ptr = ptr -> next;
        }
        return pre -> next;
    }
};